Improve graph opening times, reduce reflecion calls

dimven-adsk opened the pull request:
### Purpose

This PR targets a low-hanging fruit for improving the graph open (and saving) performance by reducing the CLR reflection calls that are used for getting a node's package assembly name (used in figuring out dependencies and used packages in a graph).

These reflection calls are not insignificant and the result can be cached (because it is not possible to change a node's assembly during runtime). If that ever changes, then you can override the method :) . Furthermore, the current implementation actually ignores the existing assembly reference and instead reloads it again from the referenced location path, which seems completely pointless.

A further improvement can be made by storing only the assembly's name as a string instead of the entire `AssemblyName` class instance, since we're only using the name.

The hot path before:
![devenv_MBWq5M7hXq](https://github.com/user-attachments/assets/922196e4-3b78-4b7d-9bff-27f7ba1a5f8c)

And after:
![devenv_oSYjkqT8f7](https://github.com/user-attachments/assets/26bfdc35-66ea-46bf-9b70-9c4ba46148a1)

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. Use N/A to indicate that the changes in this pull request do not apply to Release Notes. **Mandatory section**

### Reviewers

(FILL ME IN) Reviewer 1 (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

mjkkirschner commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -1223,18 +1223,21 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
         /// </summary>
         /// <returns>Assembly Name</returns>
-        internal virtual AssemblyName GetNameOfAssemblyReferencedByNode()
+        internal virtual string GetNameOfAssemblyReferencedByNode()
```
@dimven-adsk thanks for these changes - the cache is a good idea.
What I would prefer is that unless the change from `AssemlyName` -> `String` has clear performane benefits I would say to keep the signature of these methods the same.

I assume that the real perf gain here is from the cache and not using the GetType().Assembly  / strange double lookup / allocation of `AssemblyName`. It seems if we just cache the `AssemblyName` we'll get most of the benefit, though I guess the `AssemblyName` objects are quite wasteful at the moment though I can see us using some of these other properties in the future and I'd rather avoid spreading out the cost of allocating these later in a less predictable manner, or doing error prone string munging - thoughts?

pinzart90 replied:
I would echo the same:
`What I would prefer is that unless the change from AssemlyName -> String has clear performane benefits I would say to keep the signature of these methods the same.`

pinzart90 commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -1223,18 +1223,21 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
         }
 
+
+        internal string cachedAsmName = null;
```
I would make this protected

pinzart90 replied:
ANother name for the member could be 
`NameOfAssemblyReferencedByNode` to match the existing method

pinzart90 commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -1223,18 +1223,21 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
+            if (string.IsNullOrEmpty(cachedAsmName))
+            {
+                cachedAsmName = GetType().Assembly.FullName;
+            }
```
could probably do it 
```
NameOfAssemblyReferencedByNode ??= GetType().Assembly.GetName();
return NameOfAssemblyReferencedByNode;
```
or even
```
internal virtual AssemblyName GetNameOfAssemblyReferencedByNode() => NameOfAssemblyReferencedByNode ??= GetType().Assembly.GetName();
```

dimven-adsk commented:
I've implemented the feedback and reverted the code back to using the full AssemblyName. You are totally correct, there is no observable difference between using that and a string:

![image](https://github.com/user-attachments/assets/b2941328-5fa8-4ff0-9185-30a909e4354c)


I imagine the only difference would be a few extra bytes of memory for storing the AN instances, which would be insignificant :)
